servlet/src/main/java/io/grpc/servlet/ServletServerStream.java

```diff
@@ -0,0 +1,425 @@
+
+    final boolean trailersSent;
+
+    private WriteState(boolean stillWritePossible, boolean trailersSent) {
+      this.stillWritePossible = stillWritePossible;
+      this.trailersSent = trailersSent;
+    }
+
+    @CheckReturnValue
+    WriteState withTrailersSent(boolean trailersSent) {
+      return new WriteState(stillWritePossible, trailersSent);
+    }
+
+    /**
+     * Only {@link javax.servlet.WriteListener#onWritePossible()} can set it to true, and only
+     * {@link ServletServerStream.Sink#writeFrame} can set it to false;
+     */
+    @CheckReturnValue
+    WriteState withStillWritePossible(boolean stillWritePossible) {
+      return new WriteState(stillWritePossible, trailersSent);
+    }
+
+    @CheckReturnValue
+    WriteState newState() {
+      return new WriteState(stillWritePossible, trailersSent);
+    }
+  }
+
+  final class Sink implements AbstractServerStream.Sink {
+
+    final HttpServletResponse resp;
+
+    Sink() {
+      resp = (HttpServletResponse) asyncCtx.getResponse();
+    }
+
+    final TrailerSupplier trailerSupplier = new TrailerSupplier();
+
+    @Override
+    public void writeHeaders(Metadata headers) {
+      // Discard any application supplied duplicates of the reserved headers
+      headers.discardAll(CONTENT_TYPE_KEY);
+      headers.discardAll(GrpcUtil.TE_HEADER);
+      headers.discardAll(GrpcUtil.USER_AGENT_KEY);
+
+      if (logger.isLoggable(FINE)) {
+        logger.log(FINE, "[{0}] writeHeaders {1}", new Object[] {logId, headers});
+      }
+
+      resp.setStatus(HttpServletResponse.SC_OK);
+      resp.setContentType(CONTENT_TYPE_GRPC);
+
+      byte[][] serializedHeaders = TransportFrameUtil.toHttp2Headers(headers);
+      for (int i = 0; i < serializedHeaders.length; i += 2) {
+        resp.addHeader(
+            new String(serializedHeaders[i], StandardCharsets.US_ASCII),
+            new String(serializedHeaders[i + 1], StandardCharsets.US_ASCII));
+      }
+      resp.setTrailerFields(trailerSupplier);
+      writeFrame(FLUSH);
+    }
+
+    @Override
+    public void writeFrame(@Nullable WritableBuffer frame, boolean flush, int numMessages) {
+      if (frame == null && !flush) {
+        return;
+      }
+
+      if (logger.isLoggable(FINEST)) {
+        logger.log(
+            FINEST,
+            "[{0}] writeFrame: numBytes = {1}, flush = {2}, numMessages = {3}",
+            new Object[]{logId, frame == null ? 0 : frame.readableBytes(), flush, numMessages});
+      }
+
+      if (frame != null) {
+        int numBytes = frame.readableBytes();
+        if (numBytes > 0) {
+          onSendingBytes(numBytes);
+        }
+        writeFrame((ByteArrayWritableBuffer) frame);
+      }
+
+      if (flush) {
+        writeFrame(FLUSH);
+      }
+    }
+
+    private void writeFrame(ByteArrayWritableBuffer byteBuffer) {
+      int numBytes = byteBuffer.readableBytes();
+      WriteState curState = writeState.get();
+      if (curState.stillWritePossible) {
```

Comment:
This `stillWritePossible` logic seems complicated. It seems a lock would make this much more straight-forward.

Here it would be:
```java
synchronized (lock) {
  if (waitingOnWritePossible || !outputStream.isReady()) {
    waitingOnWritePossible = true;
    queue.offer(byteBuffer);
    return;
  }
}
// do the write
// No checking for isReady
```

In `onWritePossible`:
```java
while (true) {
  Object item;
  synchronized (lock) {
    if (!waitingOnWritePossible || queue.isEmpty()) {
      waitingOnWritePossible = false;
      return;
    }
  }
  item = queue.poll();
  // write
  if (!outputStream.isReady()) {
    return;
  }
}
```

I think the `isReady()` check within the lock is likely required, as otherwise a callback on onWritePossible could be missed. And it would be very complex to try to avoid that (which I think is the case in this code).

To check at the end:
```java
synchronized (lock) {
  if (waitingOnWritePossible) {
    queue.offer(byteBuffer);
    return;
  }
}
// do the write
synchronized (lock) {
  if (!outputStream.isReady()) {
    waitingOnWritePossible = true;
  }
}
```